sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,285 @@
+        let systemd_networkctl_result = Command::new(NETWORKCTL)
+            .arg("status")
+            .arg("--json=pretty")
+            .arg(link)
+            .output()
+            .context(error::NetworkctlExecutionSnafu)?;
+        ensure!(
+            systemd_networkctl_result.status.success(),
+            error::FailedNetworkctlSnafu {
+                stderr: String::from_utf8_lossy(&systemd_networkctl_result.stderr)
+            }
+        );
+
+        let status_output = String::from_utf8(systemd_networkctl_result.stdout)
+            .context(error::NetworkctlParsingSnafu {})?;
+
+        let networkd_status =
+            serde_json::from_str(&status_output).context(error::NetworkctlDeserializeSnafu {})?;
+
+        Ok(networkd_status)
+    }
+
+    // Fetches the IP Address for the primary interface. If there are no addresses, this is
+    // an error. If there is one, it should be returned. If there is more than one, then find
+    // the first IPv4 address and return it, if there are none, return the first IPv6 address.
+    pub(crate) fn primary_address(&self) -> Result<IpAddr> {
+        // Find IPv4 first
+        for addr in self.addresses.iter() {
+            if addr.is_ipv4() {
+                return Ok(*addr);
+            }
+        }
+        // No IPv4 addresses, then attempt to return the first IPv6, otherwise return error
+        match self.addresses.first() {
+            Some(addr) => Ok(*addr),
+            None => error::NoIpAddressSnafu {
+                interface: self.name.clone(),
+            }
+            .fail(),
+        }
+    }
+}
+
+mod error {
+    use crate::interface_id::InterfaceId;
+    use snafu::Snafu;
+    use std::{io, string::FromUtf8Error};
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(crate)))]
+    pub(crate) enum Error {
+        #[snafu(display("Failed to run 'networkctl': {}", source))]
+        NetworkctlExecution { source: io::Error },
+
+        #[snafu(display("'networkctl' failed: {}", stderr))]
+        FailedNetworkctl { stderr: String },
+
+        // Consider maybe shifting to length only for this error "Expected a Vec of length 4 but it was {}", v.len()
+        #[snafu(display("Failed to parse IP Address: {:?} {}", input, msg))]
+        BadIpAddress { input: Vec<u8>, msg: String },
+
+        #[snafu(display("Failed to parse IP Address from networkctl status"))]
+        BadIpAddressVec,
+
+        #[snafu(display("No IP Address for Primary Interface: {:?}", interface))]
+        NoIpAddress { interface: InterfaceId },
+
+        #[snafu(display("Failed to parse 'networkctl' output: {}", source))]
+        NetworkctlParsing { source: FromUtf8Error },
+
+        #[snafu(display("Failed to deserialize 'networkctl' output: {}", source))]
+        NetworkctlDeserialize { source: serde_json::Error },
+
+        #[snafu(display("No IP Address found for Primay Interface: {}", source))]
+        NoPrimaryIpAddress { source: io::Error },
```

Comment:
I don't think this error variant is used either.

Reply:
Oh, this was from my other change which will be coming next, I'll back this out and add it to the other change. Good catch.